File: nodejs/03-realtime-websocket-updates.js
const axios = require("axios");
const { io } = require("socket.io-client");

// You can generate FreeConvert API key from your user account dashboard.
// Read more at FreeConvert API reference docs: https://www.freeconvert.com/api/v1/
const apiKey = "my_api_key";

const freeconvert = axios.create({
    baseURL: "https://api.freeconvert.com/v1",
    headers: {
        "Content-Type": "application/json",
        "Accept": "application/json",
        "Authorization": `Bearer ${apiKey}`,
    },
});

async function websocketExample1() {
    console.log("websocket example 1");

    //
    // Initialize a websocket connection to FreeConvert notification service, using socket.io client library.
    // Read more: https://www.freeconvert.com/api/v1/#freeconvert-com-api-socket
    //
    const socket = io("https://notification.freeconvert.com/", {
        transports: ["websocket"],
        path: "/socket.io",
        auth: { token: `Bearer ${apiKey}` },
    });

    // Create a single task.
    const taskResponse1 = await freeconvert.post("/process/import/url", {
        url: "https://cdn.freeconvert.com/logo_theme.svg",
        filename: "logo.svg",
    });
    const taskId = taskResponse1.data.id;

    // Watch for task completion event. Following code will handle successful completion only.
    // For more events, see https://www.freeconvert.com/api/v1/?javascript#websocket-channels-amp-events
    //
    // *** Note: Events may arrive out of oder. ***
    //
    socket.on("task_completed", (data) => {
        console.log("Task completed", data.id);
        // Unsubscribe when no longer needed.
        socket.emit("unsubscribe", `task.${data.id}`);
        socket.disconnect();
    });

    // Subscribe to events of the created task.
    console.log("Start waiting for task", taskId);
    socket.emit("subscribe", `task.${taskId}`);
}

async function websocketExample2() {
    console.log("websocket example 2");

    //
    // Initialize a websocket connection to FreeConvert notification service, using socket.io client library.
    // Read more: https://www.freeconvert.com/api/v1/#freeconvert-com-api-socket
    //
    const socket = io("https://notification.freeconvert.com/", {
        transports: ["websocket"],
        path: "/socket.io",
        auth: { token: `Bearer ${apiKey}` },
    });

    // Create a job. Job will complete when all its children tasks are complete.
    const jobResponse1 = await freeconvert.post("/process/jobs", {
        tasks: {
            myImport1: {
                operation: "import/url",
                url: "https://cdn.freeconvert.com/logo_theme.svg",
                filename: "logo.svg",
            },
            myConvert1: {
                operation: "convert",
                input: "myImport1",
                output_format: "jpg",
            },
            myExport1: {
                operation: "export/url",
                input: "myConvert1",
                filename: "my-converted-file.jpg",
            },
        },
    });
    const jobId = jobResponse1.data.id;

    // Register for task and job completion events. Following code will handle successful completion only.
    // For more events, see https://www.freeconvert.com/api/v1/?javascript#websocket-channels-amp-events
    //
    // *** Note: Events may arrive out of oder. ***
    //

    socket.on("task_completed", (data) => {
        console.log("Task completed", data.name);
        // Unsubscribe when no longer needed.
        socket.emit("unsubscribe", `task.${data.id}`);
    });

    socket.on("job_completed", (data) => {
        console.log("Job completed", data.id);
        // Unsubscribe when no longer needed.
        socket.emit("unsubscribe", `job.${data.id}`);
        socket.disconnect();
    });

    // Subscribe to events of the created job.
    console.log("Start waiting for job", jobId);
    socket.emit("subscribe", `job.${jobId}`);

    // Subscribe to all tasks in the job.
    jobResponse1.data.tasks.forEach((task) => {
        socket.emit("subscribe", `task.${task.id}`);
    });
}

async function app() {
    await websocketExample1();
    await websocketExample2();
}

app().catch((error) => console.log(error.message));

File: python/03-realtime-websocket-updates.py
import os
import asyncio
import requests
import socketio

# You can generate FreeConvert API key from your user account dashboard.
# Read more at FreeConvert API reference docs: https://www.freeconvert.com/api/v1/
api_key = "my_api_key"
base_url = "https://api.freeconvert.com/v1"

freeconvert = requests.Session()
freeconvert.headers.update({
    "Content-Type": "application/json",
    "Accept": "application/json",
    "Authorization": f"Bearer {api_key}"
})

async def websocket_example1():

    print("websocket example 1")

    #
    # Initialize a websocket connection to FreeConvert notification service, using socket.io client library.
    # Read more: https://www.freeconvert.com/api/v1/#freeconvert-com-api-socket
    #
    sio = socketio.Client()
    sio.connect("https://notification.freeconvert.com", transports=["websocket"],  auth={"token": f"Bearer {api_key}"})

    # Event to signal when all is completed
    listening_finished = asyncio.Event()

    # Create a single task
    task_response = freeconvert.post(f"{base_url}/process/import/url", json={
        "url": "https://cdn.freeconvert.com/logo_theme.svg",
        "filename": "logo.svg"
    }).json()
    task_id = task_response["id"]
    

    # Watch for task completion event
    @sio.on("task_completed")
    def on_task_completed(data):
        print("Task completed", data["id"])
        # Unsubscribe when no longer needed
        sio.emit("unsubscribe", f"task.{data['id']}")
        sio.disconnect()
        listening_finished._loop.call_soon_threadsafe(listening_finished.set)

    # Subscribe to events of the created task
    print("Start waiting for task", task_id)
    sio.emit("subscribe", f"task.{task_id}")

    await listening_finished.wait()

async def websocket_example2():

    print("websocket example 2")
    
    # Initialize a websocket connection to FreeConvert notification service.
    # Create a job and watch for its completion event.
    sio = socketio.Client()
    sio.connect("https://notification.freeconvert.com", transports=["websocket"],  auth={"token": f"Bearer {api_key}"})

    # Event to signal when all is completed
    listening_finished = asyncio.Event()

    # Create a job
    job_response = freeconvert.post(f"{base_url}/process/jobs", json={
        "tasks": {
            "myImport1": {
                "operation": "import/url",
                "url": "https://cdn.freeconvert.com/logo_theme.svg",
                "filename": "logo.svg"
            },
            "myConvert1": {
                "operation": "convert",
                "input": "myImport1",
                "output_format": "jpg"
            },
            "myExport1": {
                "operation": "export/url",
                "input": "myConvert1",
                "filename": "my-converted-file.jpg"
            }
        }
    }).json()
    job_id = job_response["id"]

    # Watch for task and job completion events
    @sio.on("task_completed")
    def on_task_completed(data):
        print("Task completed", data["name"])
        # Unsubscribe when no longer needed
        sio.emit("unsubscribe", f"task.{data['id']}")

    @sio.on("job_completed")
    def on_job_completed(data):
        print("Job completed", data["id"])
        # Unsubscribe when no longer needed
        sio.emit("unsubscribe", f"job.{data['id']}")
        sio.disconnect()
        listening_finished._loop.call_soon_threadsafe(listening_finished.set)

    # Subscribe to events of the created job
    print("Start waiting for job", job_id)
    sio.emit("subscribe", f"job.{job_id}")

    # Subscribe to all tasks in the job
    for task in job_response["tasks"]:
        sio.emit("subscribe", f"task.{task['id']}")

    await listening_finished.wait()

async def app():
    await websocket_example1()
    await websocket_example2()

if __name__ == "__main__":
    try:
        asyncio.run(app())
    except Exception as e:
        print(e)
